pkg/distributor/distributor.go

```diff
@@ -1028,6 +1037,61 @@ func (d *Distributor) prePushForwardingMiddleware(next push.Func) push.Func {
 	}
 }
 
+// prePushEphemeralMiddleware is used as push.Func middleware in front of push method.
+// If marking series as ephemeral is enabled, this middleware uses the ephemeral series
+// provider to determine whether a time series should be marked as ephemeral.
+func (d *Distributor) prePushEphemeralMiddleware(next push.Func) push.Func {
+	if !d.cfg.EphemeralMetricsEnabled {
+		return next
+	}
+
+	return func(ctx context.Context, pushReq *push.Request) (*mimirpb.WriteResponse, error) {
+		userID, err := tenant.TenantID(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		req, err := pushReq.WriteRequest()
+		if err != nil {
+			return nil, err
+		}
+
+		// No ephemeral series from ruler.
+		if req.Source == mimirpb.RULE {
+			return next(ctx, pushReq)
+		}
+
+		req.EphemeralTimeseries = nil
+
+		ephemeralChecker := d.EphemeralCheckerByUser.EphemeralChecker(userID)
+		if ephemeralChecker != nil {
+			req.EphemeralTimeseries = mimirpb.PreallocTimeseriesSliceFromPool()
+
+			var deleteTs []int
+			for ix := 0; ix < len(req.Timeseries); ix++ {
+				ts := req.Timeseries[ix]
+
+				if !ephemeralChecker.IsEphemeral(ts.Labels) {
+					continue
+				}
+
+				if cap(deleteTs) == 0 {
```

Comment:
the reason why i chose `cap()` is because if `cap() == 0` then the next operation is `deleteTs = make([]int, 0, len(req.Timeseries)-ix)`, this will set the `cap()` to `> 0` but not the `len()`, so the check and the operation which is decided based on the check are closer together. Relying on `len()` would work too, but the increased `len()` is a less direct consequence because it only gets raised further down in the code.

anyway, I'm happy to change it to `len()` if you prefer that.

Reply:
This is just relying on the `first` boolean now instead of `len` or `cap`: https://github.com/grafana/mimir/pull/3897/files#diff-8e51742acf8f88fa2ae01855fb19125feb5c54c41023f1eaa8beded9f1b4bcc3R1063-R1076

This made sense because there are two separate actions that need to be taken on the first iteration, only one of which is directly related to the length of `deleteTs`.